contracts/crowdsale/validation/IndividuallyCappedCrowdsale.sol

```diff
@@ -0,0 +1,50 @@
+pragma solidity ^ 0.4.18;
+
+import "../../math/SafeMath.sol";
+import "../Crowdsale.sol";
+import "../../ownership/Ownable.sol";
+
+/**
+ * @dev Crowdsale with per-user caps
+ */
+contract IndividuallyCappedCrowdsale is Crowdsale, Ownable {
+  using SafeMath for uint256;
+
+  mapping(address => uint256) public contributions;
+  mapping(address => uint256) public caps;
+
+  /**
+   * @param _beneficiary Address to be capped
+   * @param _cap Wei limit for individual contribution
+   */
+  function setUserCap(address _beneficiary, uint256 _cap) external onlyOwner {
+    caps[_beneficiary] = _cap;
+  }
+
+  /**
+   * @param _beneficiary Address whose cap is to be checked
+   * @return Current cap for individual user
+   */
+  function getUserCap(address _beneficiary) public view returns (uint256) {
+    return caps[_beneficiary];
+  }
+
+  /**
+   * @param _beneficiary Address of contributor
+   * @return User contribution so far
+   */
+  function getUserContribution(address _beneficiary) public view returns (uint256) {
+    return contributions[_beneficiary];
+  }
+
+  function _postValidatePurchase(address _beneficiary, uint256 _weiAmount) internal {
+    super._postValidatePurchase(_beneficiary, _weiAmount);
+    require(contributions[_beneficiary] <= caps[_beneficiary]);
```

Comment:
Again, I would try to move this to `_preValidatePurchase` to save gas.

Reply:
Why moving this precondition above is going to save some gas? I mean, we will evaluate it anyway.. can you explain a little bit more?

Reply:
Because you're performing the check at the start of `buyTokens`, if you do it at the end you would have called  `_getTokenAmount`,  `_processPurchase`,  `_updatePurchasingState` and `_fowardFunds`, and any of those can make subsequent calls to another functions or even contracts. That's gassy.